firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/CRCElements.cpp

```diff
@@ -0,0 +1,391 @@
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0x527d5351;
+
+      TestCRCElementProps::PayloadBuffer output_payload;
+
+      output_payload.resize(TestCRCElementHeaderProps::header_size + input_payload.size());
+      output_payload.copy_from(
+        input_payload.buffer(), input_payload.size(), TestCRCElementHeaderProps::payload_offset);
+
+      auto computed_crc = crc_element.compute_body_crc(output_payload, crc32c);
+
+      THEN("the computed crc is equal to the expected crc") {
+        REQUIRE(computed_crc == expected_crc);
+      }
+    }
+  }
+
+  GIVEN("A crc element is initalised with payload 0x01") {
+    TestCRCElementProps::PayloadBuffer input_payload;
+    input_payload.push_back(0x01);
+
+    TestCRCElement crc_element{input_payload};
+
+    WHEN("the crc is computed from the payload") {
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0xa016d052;
+
+      TestCRCElementProps::PayloadBuffer output_payload;
+
+      output_payload.resize(TestCRCElementHeaderProps::header_size + input_payload.size());
+      output_payload.copy_from(
+        input_payload.buffer(), input_payload.size(), TestCRCElementHeaderProps::payload_offset);
+
+      auto computed_crc = crc_element.compute_body_crc(output_payload, crc32c);
+
+      THEN("the computed crc is equal to the expected crc") {
+        REQUIRE(computed_crc == expected_crc);
+      }
+    }
+  }
+
+  GIVEN("A crc element initalised with payload=[0x01, 0x02, 0x05]") {
+    TestCRCElementProps::PayloadBuffer input_payload;
+    auto data = PF::Util::make_array<uint8_t>(0x01, 0x02, 0x05);
+    for (auto& bytes : data) {
+      input_payload.push_back(bytes);
+    }
+
+    TestCRCElement crc_element{input_payload};
+    WHEN("given payload data is written to it") {
+
+      TestCRCElementProps::PayloadBuffer output_payload;
+
+      auto write_status = crc_element.write(output_payload, crc32c);
+      auto crc = crc_element.crc();
+      auto crc_compute = crc32c.compute(input_payload.buffer(), input_payload.size());
+
+      THEN("crc computed in write method is same as crcchecker compute") {
+        REQUIRE(write_status == PF::IndexStatus::ok);
+        REQUIRE(crc == crc_compute);
+      }
+    }
+  }
+
+  GIVEN("crc element is initalised with 0 bytes") {
+
+    TestCRCElementProps::PayloadBuffer input_payload;
+
+    TestCRCElement crc_element{input_payload};
+
+    WHEN("given payload data is written to it") {
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0x6D551F4C;
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+
+      auto write_status = crc_element.write(input_buffer, crc32c);
+      auto crc_payload = crc_element.crc();
+      THEN("CRC is equal to 0") {
+        REQUIRE(crc_payload == 0);
+      }
+    }
+  }
+
+  GIVEN("A CRC element is initalised with valid payload parameter") {
```

Comment:
To make the tests clearer and easier to navigate, I would suggest splitting up the scenario into four scenarios, for `compute_body_crc`, for writing, for parsing, and (assuming we still have round-tripping) for round-tripping.

In the context of parsing bodies, "valid" is not a relevant concept for the payload parameter of the CRCElement constructor - there is no way the payload can be invalid or valid, it's just a buffer which the CRCElement writes the bytes of the payload to. Actually, I'm not even sure if "valid" is a relevant concept for CRCElements at all, since the payload is just bytes and the CRCElement doesn't care what bytes are in the payload - so we should not be using "valid" to talk about payloads in this file. You could instead write "Given a CRC Element initialized with a non-empty payload buffer with a capacity of 254 bytes"; we also need to have some tests with "Given a CRC Element initialized with an empty payload buffer with a capacity of 254 bytes"

Reply:
Added scenarios for each as suggested, and updated the given for this situation and other applicable cases.